### PR TITLE
[MIRROR] [READY] Goon Genetics

### PR DESCRIPTION
**Pull request by ghost**
Original PR: https://github.com/tgstation/tgstation/pull/41258
----------------------------
### GENETICS

- The random hexadecimal rng game has been replaced with gene sequencing from goon.

- Adds mutation activators and mutators

- You can now store mutations

- Everyone now has their own set of unique mutations

- Limited mutations per person to 8 (including one always being monkey)

- Adds race specific mutations (See fire breathing for lizads)

- You can inspect discovered mutations, undiscovered mutations use an alias to recognize them by

- Adds a sequence analyzer. Can be used to scan someones genes sequence. Useful for determing what mutations they can safely have and or collecting data for very difficult sequences

- Adds mutation combining. It's currently only RADIOACTIVE + STRONG = HULK  (So yes you will now need 2 mutations for to get hulk)

- Adds several other mutations. Telepathy, firebreath, glowy, radioactive and strength

:cl: Time-Green
add: Goon genetics!
add: More mutations! Fire breath for lizards! Radioactive! Telepathy! Glowy! Strength, though its cosmetic and should be combined with radioactivity instead! Fiery sweat! 
add: Adds void magnet mutation by @tralezab !
/:cl:

Quick guide on how to have fun with mutations:

Every mutation consists of a sequence. It's the AT GC stuff. A goes with T and G goes with C. An active mutation has a correct sequence, like here:
image

First we want to disable this mutation, so we can move on with the others. Do this by making it faulty.
image
We replaced the fourth A with an X, making it a wrong sequence and thus disabling the mutation.

Now we can start with the real stuff. We'll select another mutation.
image
There's four X's, making it wrong. We can instantly fix two X's. The first pair with GX should be GC. The third pair with XA should be TA.

Now our mutation should look like this:
image
That just leaves a pair of X's. There's only four possibilities: AT, TA, GC and CG

image
It was CG. The mutation is now active and can be properly toyed with. 
Note: With two X pairs, there is still only 16 possibilities. If there are even more, you might consider peaking at someone else with the same mutation to see if they got that pair. (I recommend you do this with the sequence analyzer in your pocket, wich you can click on the DNA console to update it with the latest mutation names)

If we saved the mutation, it goes to our storage (Mutations)
image

Here we can do a bunch of stuff with it. 
Mutation name: Shows you more info
image
Export: Export the mutation to a disk for storage
Delete: Removes the mutation from your storage
Combine: First click, selects the mutation. Second click, attempts to merge it with another mutation to create a new one. Currently only works with Strength and Radioactive

Other notes about UI:
Removed rejuvenators. Get some from the chemist or use antitoxins. Note that most radiation damage has been reduced by alot.
Scramble DNA: Don't like the your subjects mutations and can't get a new one? Absolutely fucking radiation blast them and generate a new set of mutations
Disk: Disk interaction. Import, export and disk ejecting fall under this
Genetic Sequencer: Where you will be spending most of your time
Unique Identifiers: All identity mutation stuff has been moved to there. It still features buffers and the UE gene pulser for all your identity stealing needs.

Activators! Already got a mutation and want to give it to someone with it in their potential? Ram this thing into them to safely give them that mutation.
Mutators! Want to give someone a mutation that isn't listed on them? Use these to instantly give it to them for nothing*! *(actually violent humiliation and death)

Featuring more mutations!
Firebreath!
Telepathy!
Glowy!
Radioactive!
Strength! (Doesnt do anything)
Fiery sweat! 
Void magnet! Made by @tralezab 
And even some old ones!
Also disabled the unstable dna mutation, since it doesn’t fit well with goon genetics

End of guide 

- [X] Make mutations random for every person
- [X] New html
- [X] Make mutations stable and working
- [x] Add mutation disks
- [X] Mutation activators
- [X] Mutation injectors
- [X] More mutations
- [x] Even more mutations
- [x] Have genetic instability do something
- [x] Add some form of viewing your own DNA for discovered mutations
- [X] Alot of testing
- [x] Add the goon gene minigame

I will also be testing everything thoroughly and tweaking some stuff.

Another PR:
Chromosomes
Even more than more mutations! (Probably by someone else since I'll be prioritizing chromosomes)

### CODE REVIEW

**Comment by ghost** [comment]
Heck yeah

**Comment by WarRage333** [comment]
idea is good but 30 minutes to get all super powers on hippie...

**Comment by GuyonBroadway** [comment]
It takes 30 mins because there's a shittone more powers.

File: code/__DEFINES/DNA.dm [hunk 1]
```diff
@@ -56,7 +66,8 @@
 #define DNA_FACIAL_HAIR_STYLE_BLOCK	6
 #define DNA_HAIR_STYLE_BLOCK		7
 
-#define DNA_STRUC_ENZYMES_BLOCKS	18
+#define DNA_SEQUENCE_LENGTH			4
+#define DNA_MUTATION_BLOCKS			8
 #define DNA_UNIQUE_ENZYMES_LEN		32
 
 //Transformation proc stuff
```
Comment (yoyobatty) [inline code/__DEFINES/DNA.dm:73]: Good.

File: code/__HELPERS/dna.dm [hunk 2]
```diff
@@ -0,0 +1,8 @@
+//////////////////////////////////////////////////////////
+//A bunch of helpers to make genetics less of a headache//
+//////////////////////////////////////////////////////////
+
+#define get_initialized_mutation(A) GLOB.all_mutations[A]
+#define mutation_in_sequence(A, B) ((A) in B.mutation_index)
+#define get_gene_string(A, B) (B.mutation_index[A])
+#define get_sequence(A) (GLOB.full_sequences[A])
```
Comment (yoyobatty) [inline code/__HELPERS/dna.dm:8]: Good

File: code/_globalvars/genetics.dm [hunk 3]
```diff
@@ -21,8 +21,10 @@ GLOBAL_LIST(op_se_blocks)
...
-
+GLOBAL_LIST_EMPTY(all_mutations)
+GLOBAL_LIST_EMPTY(full_sequences)
 GLOBAL_LIST_EMPTY(bad_mutations)
 GLOBAL_LIST_EMPTY(good_mutations)
-GLOBAL_LIST_EMPTY(not_good_mutations)
\ No newline at end of file
+GLOBAL_LIST_EMPTY(not_good_mutations)
+
+GLOBAL_LIST_EMPTY(mutation_recipes)
```
Comment (yoyobatty) [inline code/_globalvars/genetics.dm:30]: Good

File: code/_onclick/other_mobs.dm [hunk 4]
```diff
@@ -93,7 +93,7 @@
 			return
 
 	for(var/datum/mutation/human/HM in dna.mutations)
-		HM.on_ranged_attack(src, A, mouseparams)
+		HM.on_ranged_attack(A, mouseparams)
 
 	if(isturf(A) && get_dist(src,A) <= 1)
 		src.Move_Pulled(A)
```
Comment (yoyobatty) [inline code/_onclick/other_mobs.dm:99]: Good

File: code/controllers/subsystem/atoms.dm [hunk 5]
```diff
@@ -108,16 +108,19 @@ SUBSYSTEM_DEF(atoms)
...
+		var/datum/mutation/human/B = new path ()
+		B.alias = "Mutation #[i]"
+		GLOB.all_mutations[B.type] = B
+		GLOB.full_sequences[B.type] = generate_gene_sequence(B.blocks)
+		if(B.locked)
 			continue
-		B.dna_block = pick_n_take(avnums)
 		if(B.quality == POSITIVE)
 			GLOB.good_mutations |= B
 		else if(B.quality == NEGATIVE)
```
Comment (yoyobatty) [inline code/controllers/subsystem/atoms.dm:126]: Good

File: code/datums/brain_damage/mild.dm [hunk 6]
```diff
@@ -73,7 +73,7 @@
 
 //no fiddling with genetics to get out of this one
 /datum/brain_trauma/mild/speech_impediment/on_life()
-	if(!(GLOB.mutations_list[UNINTELLIGIBLE] in owner.dna.mutations))
+	if(!(owner.dna.check_mutation(UNINTELLIGIBLE)))
 		on_gain()
 	..()
```
Comment (yoyobatty) [inline code/datums/brain_damage/mild.dm:79]: Good

File: code/datums/components/spill.dm [hunk 7]
```diff
@@ -1,56 +1,56 @@
...
+/datum/component/spill/proc/drop_react(obj/item/source, mob/dropper)
+	UnregisterSignal(dropper, COMSIG_LIVING_STATUS_KNOCKDOWN)
+
+/datum/component/spill/proc/knockdown_react(mob/living/fool)
+	var/obj/item/master = parent
+	fool.dropItemToGround(master)
+	if(droptext)
+		fool.visible_message(arglist(droptext))
+	if(dropsound)
+		playsound(master, pick(dropsound), 30)
```
Comment (yoyobatty) [inline code/datums/components/spill.dm:56]: Good

File: code/datums/datacore.dm [hunk 8]
```diff
@@ -277,7 +277,6 @@
 		L.fields["sex"]			= H.gender
 		L.fields["blood_type"]	= H.dna.blood_type
 		L.fields["b_dna"]		= H.dna.unique_enzymes
-		L.fields["enzymes"]		= H.dna.struc_enzymes
 		L.fields["identity"]	= H.dna.uni_identity
 		L.fields["species"]		= H.dna.species.type
 		L.fields["features"]	= H.dna.features
```
Comment (yoyobatty) [inline code/datums/datacore.dm:282]: Good

File: code/datums/diseases/advance/symptoms/genetics.dm [hunk 9]
```diff
@@ -74,5 +74,5 @@ Bonus
 		if(M && archived_dna)
 			if(!M.has_dna())
 				return
-			M.dna.struc_enzymes = archived_dna
+			M.dna.mutation_index = archived_dna
 			M.domutcheck()
```
Comment (yoyobatty) [inline code/datums/diseases/advance/symptoms/genetics.dm:78]: Good

File: code/datums/dna.dm [hunk 10]
```diff
@@ -385,9 +543,9 @@
...
 		return 0
 	if(se)
-		for(var/i=1, i<=DNA_STRUC_ENZYMES_BLOCKS, i++)
+		for(var/i=1, i<=DNA_MUTATION_BLOCKS, i++)
 			if(prob(probability))
-				M.dna.struc_enzymes = setblock(M.dna.struc_enzymes, i, random_string(DNA_BLOCK_SIZE, GLOB.hex_characters))
+				M.dna.generate_dna_blocks()
 		M.domutcheck()
 	if(ui)
 		for(var/i=1, i<=DNA_UNI_IDENTITY_BLOCKS, i++)
```
Comment (yoyobatty) [inline code/datums/dna.dm:551]: Good

File: code/datums/footstep.dm [hunk 11]
```diff
@@ -1,99 +1,99 @@
...
+				playsound(T, pick(GLOB.footstep[T.footstep][1]),
+					GLOB.footstep[T.footstep][2] * v,
+					TRUE,
+					GLOB.footstep[T.footstep][3] + e)
+
+			if((!H.shoes && !feetCover) || !(H.mobility_flags & MOBILITY_STAND)) //are we NOT wearing shoes or are we lying/crawling (using hands to move around)?
+				playsound(T, pick(GLOB.barefootstep[T.barefootstep][1]),
+					GLOB.barefootstep[T.barefootstep][2] * v,
+					TRUE,
+					GLOB.barefootstep[T.barefootstep][3] + e)
```
Comment (yoyobatty) [inline code/datums/footstep.dm:99]: Good

File: code/datums/mutations.dm [hunk 12]
```diff
@@ -118,17 +112,26 @@ GLOBAL_LIST_EMPTY(mutations_list)
...
 				apply_overlay(CM.layer_used)
+
+/datum/mutation/human/proc/copy_mutation(datum/mutation/human/HM) //Not yet implemented, useful for when assigning specific stats.
+
+/datum/mutation/human/proc/remove()
+	if(dna)
+		dna.force_lose(src)
+	else
+		qdel(src)
+
```
Comment (yoyobatty) [inline code/datums/mutations.dm:137]: Good

File: code/datums/mutations/body.dm [hunk 13]
```diff
@@ -147,3 +160,57 @@
...
+	owner.physiology.burn_mod *= 0.5
+
+/datum/mutation/human/fire/on_losing(mob/living/carbon/human/owner)
+	if(..())
+		return
+	owner.physiology.burn_mod *= 2
+
+
+
+
```
Comment (yoyobatty) [inline code/datums/mutations/body.dm:216]: Questioning the tg logic of the fire mutation acquire and lose logic but whatever, it's good.

File: code/datums/mutations/chameleon.dm [hunk 14]
```diff
@@ -1,25 +1,26 @@
...
 
-/datum/mutation/human/chameleon/on_move(mob/living/carbon/human/owner)
+/datum/mutation/human/chameleon/on_move()
 	owner.alpha = CHAMELEON_MUTATION_DEFAULT_TRANSPARENCY
 
-/datum/mutation/human/chameleon/on_attack_hand(mob/living/carbon/human/owner, atom/target, proximity)
+/datum/mutation/human/chameleon/on_attack_hand(atom/target, proximity)
 	if(proximity) //stops tk from breaking chameleon
 		owner.alpha = CHAMELEON_MUTATION_DEFAULT_TRANSPARENCY
 		return
```
Comment (yoyobatty) [inline code/datums/mutations/chameleon.dm:26]: Good

File: code/datums/mutations/cold_resistance.dm [hunk 15]
```diff
@@ -1,32 +1,30 @@
...
+	owner.add_trait(TRAIT_RESISTLOWPRESSURE, "space_adaptation")
 
-/datum/mutation/human/cold_resistance/on_losing(mob/living/carbon/human/owner)
+/datum/mutation/human/space_adaptation/on_losing(mob/living/carbon/human/owner)
 	if(..())
 		return
-	owner.remove_trait(TRAIT_RESISTCOLD, "cold_resistance")
-	owner.remove_trait(TRAIT_RESISTLOWPRESSURE, "cold_resistance")
+	owner.remove_trait(TRAIT_RESISTCOLD, "space_adaptation")
+	owner.remove_trait(TRAIT_RESISTLOWPRESSURE, "space_adaptation")
```
Comment (yoyobatty) [inline code/datums/mutations/cold_resistance.dm:30]: Good

File: code/datums/mutations/combined.dm [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+		return FALSE
+	if(mutation1 == mutation2) //this could otherwise be bad
+		return FALSE
+	for(var/A in GLOB.mutation_recipes)
+		if(findtext(A, "[mutation1]") && findtext(A, "[mutation2]"))
+			return GLOB.mutation_recipes[A]
+
+/datum/generecipe/hulk
+	required = "/datum/mutation/human/strong; /datum/mutation/human/radioactive"
+	result = HULK
```
Comment (yoyobatty) [inline code/datums/mutations/combined.dm:16]: Good

File: code/datums/mutations/hulk.dm [hunk 17]
```diff
@@ -16,11 +19,11 @@
...
-/datum/mutation/human/hulk/on_attack_hand(mob/living/carbon/human/owner, atom/target, proximity)
+/datum/mutation/human/hulk/on_attack_hand(atom/target, proximity)
 	if(proximity) //no telekinetic hulk attack
 		return target.attack_hulk(owner)
 
-/datum/mutation/human/hulk/on_life(mob/living/carbon/human/owner)
+/datum/mutation/human/hulk/on_life()
 	if(owner.health < 0)
 		on_losing(owner)
 		to_chat(owner, "<span class='danger'>You suddenly feel very weak.</span>")
```
Comment (yoyobatty) [inline code/datums/mutations/hulk.dm:29]: Good

File: code/datums/mutations/radioactive.dm [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+/datum/mutation/human/radioactive/on_life()
+	radiation_pulse(owner, 20)
+
+/datum/mutation/human/radioactive/New()
+	..()
+	if(!(type in visual_indicators))
+		visual_indicators[type] = list(mutable_appearance('icons/effects/genetics.dmi', "radiation", -MUTATIONS_LAYER))
+
+/datum/mutation/human/radioactive/get_visual_indicator()
+	return visual_indicators[type][1]
```
Comment (yoyobatty) [inline code/datums/mutations/radioactive.dm:19]: Good

File: code/datums/mutations/speech.dm [hunk 19]
```diff
@@ -216,8 +220,9 @@
 
 /datum/mutation/human/stoner
 	name = "Stoner"
+	desc = "A common mutation that severely decreases intelligence."
 	quality = NEGATIVE
-	dna_block = NON_SCANNABLE
+	locked = TRUE
 	text_gain_indication = "<span class='notice'>You feel...totally chill, man!</span>"
 	text_lose_indication = "<span class='notice'>You feel like you have a better sense of time.</span>"
```
Comment (yoyobatty) [inline code/datums/mutations/speech.dm:228]: Good

File: code/datums/mutations/telekinesis.dm [hunk 20]
```diff
@@ -1,18 +1,20 @@
...
+		visual_indicators[type] = list(mutable_appearance('icons/effects/genetics.dmi', "telekinesishead", -MUTATIONS_LAYER))
 
-/datum/mutation/human/telekinesis/get_visual_indicator(mob/living/carbon/human/owner)
-	return visual_indicators[1]
+/datum/mutation/human/telekinesis/get_visual_indicator()
+	return visual_indicators[type][1]
 
-/datum/mutation/human/telekinesis/on_ranged_attack(mob/living/carbon/human/owner, atom/target)
+/datum/mutation/human/telekinesis/on_ranged_attack(atom/target)
 	target.attack_tk(owner)
```
Comment (yoyobatty) [inline code/datums/mutations/telekinesis.dm:20]: Good

File: code/datums/weather/weather_types/radiation_storm.dm [hunk 21]
```diff
@@ -38,9 +38,9 @@
 					H.randmuti()
 					if(prob(50))
 						if(prob(90))
-							H.randmutb()
+							H.easy_randmut(NEGATIVE+MINOR_NEGATIVE)
 						else
-							H.randmutg()
+							H.easy_randmut(POSITIVE)
 						H.domutcheck()
 		L.rad_act(20)
```
Comment (yoyobatty) [inline code/datums/weather/weather_types/radiation_storm.dm:46]: GOod

File: code/game/machinery/cloning.dm [hunk 22]
```diff
@@ -170,18 +172,18 @@
...
-		H.randmutvg()
+		H.easy_randmut(POSITIVE)
 	if(efficiency < 3)
 		if(prob(50))
 			H.gain_trauma_type(BRAIN_TRAUMA_MILD, TRAUMA_RESILIENCE_BASIC)
 		if(prob(50))
-			var/mob/M = H.randmutb()
+			var/mob/M = H.easy_randmut(NEGATIVE+MINOR_NEGATIVE)
 			if(ismob(M))
 				H = M
```
Comment (yoyobatty) [inline code/game/machinery/cloning.dm:189]: Good but needs to be mirrored to modular file.

**Review by yoyobatty — APPROVED** [review]

**Comment by yoyobatty** [comment]
I hate my job so much

**Pull request closed without merging**